Fixed #22 Propose new pyfftw_sdp

NimaSarajpoor opened the pull request:
This PR addresses issue #22. The proposed pyfftw-based sdp will first be added under `challenger_sdp.py` so that we can compare it against the existing  [pyfftw_sdp](https://github.com/stumpy-dev/sliding_dot_product/blob/main/sdp/pyfftw_sdp.py). Eventually, once we are certain that there is no other concern, we will move it to `pyfftw_sdp`.

NimaSarajpoor committed:
propose revised pyfftw in challenger

NimaSarajpoor commented on timing.py:
```diff
@@ -59,7 +59,7 @@
     p_diff = args.pdiff
 
-    if not noheader:
+    if noheader:
```
@seanlaw 
I've included this minor fix in this PR.

NimaSarajpoor replied:
Please ignore

NimaSarajpoor committed:
minor fix

NimaSarajpoor commented:
**timing.sh:**
```
#!/bin/bash

rm -rf sdp/__pycache__
./timing.py -timeout 5.0 -pmin 2 -pmax 24 -pdiff 100 pyfftw challenger > timing.csv
rm -rf sdp/__pycache__
```


**The performance of challenger relative to existing `pyffte_sdp`**

<img width="1372" height="3759" alt="challenger-performance" src="https://github.com/user-attachments/assets/e16eb230-a664-4656-9d50-cab4383f894c" />

**Observation**
The proposed challenger is approximately `1.25x-2x` faster than `pyfftw_sdp` for large `T`, where `len(T) >= 2^15`.

NimaSarajpoor commented:
@seanlaw 
If there is no particular concern on the logic, please let me know and I can move the script to `pyfftw_sdp`.

seanlaw commented:
@NimaSarajpoor Please go ahead!

NimaSarajpoor committed:
update files

NimaSarajpoor committed:
improve comments

NimaSarajpoor committed:
fix black formating

NimaSarajpoor commented:
The script for `pyfftw_sdp` has been updated to reflect the recent proposal. Given this change to our baseline performance, it is worth it to re-evaluate the performance of other modules relative to this new baseline. The following figure illustrates the performance of `pocketfft_r2c_c2r` (blue plot) and `scipy_oaconvolve` (orange plot) benchmarked against the new baseline: [pyfftw_sdp](https://github.com/stumpy-dev/sliding_dot_product/blob/challenge_pyfftw/sdp/pyfftw_sdp.py).



```
#!/bin/bash

rm -rf sdp/__pycache__
./timing.py -timeout 5.0 -pmin 6 -pmax 24 pyfftw pocketfft_r2c_c2r scipy_oaconvolve > timing.csv
rm -rf sdp/__pycache__
```

<img width="1372" height="2835" alt="pyfftw-vs-others-sdp" src="https://github.com/user-attachments/assets/26c4afce-c4f3-4fa2-8eca-85c0126873a1" />


**Observations**
* In all cases, `pyfftw_sdp` outperforms `pocketfft_r2c_c2r_sdp` 
* For `len(Q)<2^15`, `scipy_oaconvolve_sdp` outperforms (single-threaded) `pyfftw_sdp` when `len(T) >> len(Q)`.

**Conclusion**
For `len(Q) >= 2^6`, our focus can be on the following two modules:
* `scipy_oaconvolve_sdp`
* `pyfftw_sdp`

I think it should be worth it to check the performance of multi-threading `pyfftw_sdp` and see if that outperforms `scipy_oaconvolve_sdp`.

NimaSarajpoor commented:
> I think it is worth evaluating the performance of multi-threaded `pyfftw_sdp` to see whether it can outperform `scipy_oaconvolve_sdp`.

The baseline implementation of `pyfftw_sdp` is single-threaded. In what follows, I explore multi-threaded variants of `pyfftw_sdp`, where the number of threads (`self.threads`, see the line below) is set to 2, 4, and 8:

https://github.com/stumpy-dev/sliding_dot_product/blob/d8f0423030b23b3860300a4ebeb2688908853e16/sdp/pyfftw_sdp.py#L11

**The benchmark was run using the following `timing.sh` script:**

```bash
#!/bin/bash

rm -rf sdp/__pycache__
./timing.py -timeout 5.0 -pmin 6 -pmax 24 pyfftw challenger_2threads challenger_4threads challenger_8threads scipy_oaconvolve > timing.csv
rm -rf sdp/__pycache__
```

**Results**

<img width="1354" height="2835" alt="pyfftwThreads-vs-oaconvolve" src="https://github.com/user-attachments/assets/a0b57afd-38c3-478f-b1d1-ca04ffd8c708" />

**Observations**
* Multi-threaded `pyfftw_sdp` does not consistently outperform the single-threaded baseline. In fact, for shorter input arrays, the single-threaded version often performs better.
* `pyfftw_sdp` with 4 or 8 threads outperforms `scipy_oaconvolve`.

**Side note**
It is important to clarify a key assumption underlying this comparison: the FFT planning time in pyfftw is excluded. This assumes that the FFT is applied repeatedly to inputs of the same size, in which case the one-time planning cost (which can also be computed in advance) can reasonably be ignored.
Under this assumption, single-threaded / multi-threaded `pyfftw_sdp` seems to be the solution we are looking for. However, if the application requires FFTs on new, unforeseen large input sizes, then `scipy_oaconvolve` may be the better choice, as it does not have to spend time for the "planning" phase.

NimaSarajpoor commented:
@seanlaw 
Do you think this PR is ready to be merged?

seanlaw commented:
> Do you think this PR is ready to be merged?

Please give me some time to review it more thoroughly

seanlaw commented on sdp/pyfftw_sdp.py:
```diff
@@ -7,36 +7,64 @@ class SLIDING_DOT_PRODUCT:
     def __init__(self):
         self.m = 0
```
Is `self.m` even needed? 

It looks like it's only being used internally in `__call__`

NimaSarajpoor replied:
You are right! Will remove the attribute.

seanlaw commented on sdp/pyfftw_sdp.py:
```diff
@@ -7,36 +7,64 @@ class SLIDING_DOT_PRODUCT:
         self.m = 0
         self.n = 0
+        self.shape = 0
         self.threads = 1
```
Instead of `self.threads`, would it make sense to make `threads=1` a parameter for `__call__`?

NimaSarajpoor replied:
It makes sense! Do you think we should add `planning_flag='FFTW_MEASURE'` to `__call__` as well for the sake of completeness?

seanlaw replied:
Yeah, that would be better

seanlaw commented on sdp/pyfftw_sdp.py:
```diff
@@ -7,36 +7,64 @@ class SLIDING_DOT_PRODUCT:
+
+        if need_new_obj:
+            # create input and output arrays and FFTW objects
+            self.real_arr = pyfftw.empty_aligned(self.shape, dtype="float64")
```
I think that we should always instantiate class attributes in `__init__`. 

Having said that, instead of needing to check the shape, is there an opportunity to make `self.real_arr` a dictionary and using `shape` as the key? Then, when the array (with the desired shape) doesn't exist, then you can create the array on the fly?

Here's a simple illustrative example of this dict
```
self.real_arr  = {}
shape = 1024
real_arr = self.real_arr.get(shape, pyfftw.empty_aligned(shape, dtype="float64"))
```

This way, the array is also being stored for later use

seanlaw replied:
I guess that this can be applied for `complex_arr` and the `rfft_obj` and `irfft_obj` as well

seanlaw replied:
It's possible that I cam overcomplicating things too...

NimaSarajpoor replied:
@seanlaw 
> I think that we should always instantiate class attributes in __init__.

👍 

> Having said that, instead of needing to check the shape, is there an opportunity to make self.real_arr a dictionary and using shape as the key? Then, when the array (with the desired shape) doesn't exist, then you can create the array on the fly?

> I guess that this can be applied for complex_arr and the rfft_obj and irfft_obj as well

We can definitely add that mechanism here for `rfft_obj` and `irfft_obj`. 

For `real_arr` and `complex_arr`, I previously tried a different approach and was wondering whether we should apply it here. The idea was to preallocate a large `real_arr` in advance, using a length of `2^20`, although this could be user-configurable. If the required shape is smaller, we simply take a slice of this preallocated array. If the shape exceeds 2^20, we create a new array on the fly, use it, and then discard it. On second thought, what do you think about combining this approach with the dictionary-based approach for managing the arrays? We allow user to create a pre-allocated large arrays for both real and complex. Then, we can just put a slice of it (a view) to dictionary. If it is bigger than that, we create new array and then store it in dictionary.

seanlaw replied:
> The idea was to preallocate a large real_arr in advance, using a length of 2^20, although this could be user-configurable. If the required shape is smaller, we simply take a slice of this preallocated array. If the shape exceeds 2^20, we create a new array on the fly, use it, and then discard it.

I didn't realize that it could be re-used for smaller arrays. This changes my earlier comments! You should start with an average sized array by default and then, when you need something longer, create it and replace the existing (shorter) one for the longer array. You can allow the user to specify the starting array size but have 2^20 (or whatever as default). You wouldn't need a dict in this case. Instead, you'd just check if you needed to expand the array to accommodate.

seanlaw reviewed (commented):
@NimaSarajpoor I basically have one primary suggestion for you to consider but I am not married to it (see below)

NimaSarajpoor committed:
addressed comments

NimaSarajpoor committed:
fixed unnecessary import

NimaSarajpoor commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,48 +4,86 @@
     # https://stackoverflow.com/a/30615425/2955541
-    def __init__(self):
-        self.m = 0
+    def __init__(self, max_shape=2**10):
```
Should we remove `max_shape` from here, and get its value from `config.py` instead? 

```
def __init__(self):
    max_shape = config.PYFFTW_MAX_SHAPE
```

seanlaw replied:
No, it should be kept here as an initialization parameter. 

I don't really like using the word "shape" because "shape" almost implies multiple dimensions. Here, we really mean `len(T)`, which is typically `n`. Therefore, `max_n` makes more sense (from a STUMPY standpoint).

I understand that `pyfftw` is using the word "shape" but I think we should be more specific here.

NimaSarajpoor commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,48 +4,86 @@
+        self.shape = 0
+        self.real_arr = pyfftw.empty_aligned(max_shape, dtype="float64")
+        self.complex_arr = pyfftw.empty_aligned(1 + max_shape // 2, dtype="complex128")
+        self.rfft_irfft_objects = {}
```
Need to add a comment here and explain that key is `shape`, and value is tuple `(rfft_object, irrft_object)`. 

----

Should we break it down to two objects: `self.rfft_objects={}` and `self.irfft_objects={}` ?

seanlaw replied:
Two objects is probably better

NimaSarajpoor commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,48 +4,86 @@
 
 
-_sliding_dot_product = SLIDING_DOT_PRODUCT()
+_sliding_dot_product = SLIDING_DOT_PRODUCT(max_shape=2**10)
```
`max_shape` is set to `2**10` so that the condition `if self.shape > len(self.real_arr):` becomes true at some point in the test function `test_sdp_power2` so that we get 100% coverage. 

However, I think maybe we should create `config.py` and add variable `PYFFTW_MAX_SHAPE` to it. Then, in the test function, we can create a particular test to make sure   `if self.shape > len(self.real_arr):` is triggered.

seanlaw replied:
It feels like we're overcomplicating things. Instead, you can create a simple test where you initialize `SLIDING_DOT_PRODUCT(max_shape=2**6)` and then pass in an array that is clearly longer than `2**6` and this will ensure that `if self.shape > len(self.real_arr):` gets triggered.

NimaSarajpoor replied:
> Instead, you can create a simple test where you initialize SLIDING_DOT_PRODUCT(max_shape=2**6) 

Right!! 👍

NimaSarajpoor commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,48 +4,86 @@
+        ----------
+        max_shape : int
+            Maximum shape to preallocate arrays for. This will be the size of the
+            the real-valued array. A complex-valued array of size 1 + max_shape // 2
```
```suggestion
            real-valued array. A complex-valued array of size 1 + max_shape // 2
```

seanlaw commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,76 +4,86 @@
+        ----------
+        max_shape : int
+            Maximum shape to preallocate arrays for. This will be the size of the
+            the real-valued array. A complex-valued array of size 1 + max_shape // 2
```
Is it `(1 + max_shape) // 2` or `1 + (max_shape // 2)`? Please add parentheses where appropriate

NimaSarajpoor replied:
The latter... will add parentheses to improve the clarity

seanlaw commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,76 +4,86 @@
-            if self.shape != shape:
-                self.shape = shape
-                need_new_obj = True
+            self.shape = pyfftw.next_fast_len(self.n)
```
Honestly, instead of using `self.shape`, I would rather use `self.next_fast_n` or something "better" like that. Every time I see `shape`, my brain thinks that there are other dimensions and that it matters

seanlaw reviewed (commented):
@NimaSarajpoor I've left some comments for you to consider

NimaSarajpoor committed:
add comment to improve clarity

NimaSarajpoor committed:
addressed comments

NimaSarajpoor committed:
increase the default value of max_n parameter

NimaSarajpoor committed:
added test function

NimaSarajpoor committed:
renamed test function

NimaSarajpoor committed:
fixed minor issue

NimaSarajpoor committed:
minor change

NimaSarajpoor committed:
Revised comment

NimaSarajpoor commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,48 +4,99 @@
-            shape = pyfftw.next_fast_len(self.n)
-            self.rfft_Q_obj = pyfftw.builders.rfft(
-                np.empty(self.m), overwrite_input=True, n=shape, threads=self.threads
+            self.next_fast_n = pyfftw.next_fast_len(self.n)
```
@seanlaw 
I decided to use the proposed name `next_fast_n` since this is close to the function name `next_fast_len`.

NimaSarajpoor commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,48 +4,99 @@
+        self.real_arr = pyfftw.empty_aligned(max_n, dtype="float64")
+        self.complex_arr = pyfftw.empty_aligned(1 + (max_n // 2), dtype="complex128")
+
+        # Store FFTW objects in a dict keyed by `next_fast_n`, where n=len(T)
```
@seanlaw 
Note that, in `irfft`, the length of OUTPUT is `next_fast_n` (NOT the length of INPUT). However, since computing `SDP(Q, T)` requires both `rfft` and `irfft` objects, and `next_fast_n` is the length based on which the length of input/output arrays are determined, I am using the key `next_fast_n` for both objects. 

---

**As an aside**:

Should the dictionary key be the tuple `(next_fast_n, threads, planning_flag)` rather than just `next_fast_n`?

Currently, if a user attempts to benchmark the impact of different `threads` or `planning_flag` settings for a fixed input length, the dictionary will only store the first rfft and irfft objects created. Because the key is limited to `next_fast_n`, subsequent changes to the other parameters won't trigger the creation of new objects.

seanlaw replied:
> Should the dictionary key be the tuple (next_fast_n, threads, planning_flag) rather than just next_fast_n?

1. I suggest that we use `n_threads` instead of `threads`
2. Is it actually possible to use `planning_flag` as a part of the tuple? If so, having all three in a tuple is fine

NimaSarajpoor replied:
> I suggest that we use `n_threads` instead of `threads`

It makes sense. Will do.

> Is it actually possible to use `planning_flag` as a part of the tuple? 

The value provided for `planning_flag` has the `string` type. It should be okay to use it as part of the tuple. I will improve docstring to explain that the following values are supported for `planning_flag`: 
*  `FFTW_ESTIMATE`
* `FFTW_MEASURE`
* `FFTW_PATIENT` 
* `FFTW_EXHAUSTIVE`

NimaSarajpoor committed:
addressed comments

NimaSarajpoor committed:
minor changes

NimaSarajpoor committed:
minor change in docstring

NimaSarajpoor committed:
minor change in docstring

NimaSarajpoor committed:
add comment to test function to show its purpose

NimaSarajpoor commented:
@seanlaw 
I've addressed the comments. I've also improved the comments and docstrings in `pyfftw_sdp.py`.  Additionally, improved the comment in the new test function that was added recently to `test.py`.

seanlaw commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,48 +4,125 @@
+            Sliding dot product between `Q` and `T`.
+        """
+        m = Q.shape[0]
+        if self.n != T.shape[0]:
```
I don't think this if statement is needed. The lines below can still be executed regardless of whether the length has changed

NimaSarajpoor replied:
Right..... the only reason for having that if statement is to avoid calling `pyfftw.next_fast_len` when possible. It should be fine though to remove it as that results in just a slight drop in the performance for lengths `2^18` to `2^21`  (0%-10%).

seanlaw reviewed (commented):
@NimaSarajpoor I think this is ready to be merged. I left a comment for you to consider but feel free to ignore if you disagree as I do not feel strongly about it

NimaSarajpoor committed:
avoid tracking len(T) via attribute

NimaSarajpoor commented on sdp/pyfftw_sdp.py:
```diff
@@ -4,48 +4,120 @@
-                threads=self.threads,
+            self.rfft_objects[key] = rfft_obj
+        else:
+            rfft_obj.update_arrays(real_arr, complex_arr)
```
**An opportunity for (minor) performance improvement.**
In this line, the method `.update_arrays` is called unless the rfft object is created for the first time . HOWEVER, this `.update_arrays` is actually needed only when the pre-allocated arrays that were used initially during the creation of the rfft object are (destroyed and) replaced with longer arrays. 


**How to implement?**
We can track the `id` of the pre-allocated array. And whenever an object is fetched from the dictionary, we can compare the current id with the `id` of the array that was used before. 

```
def __init__():
    # add attribute: `self.array_ids = {}`  # to store `id` of self.real_arr

def __call__(...):
    # some code

    current_id = id(self.real_arr)   # id of pre-allocated array in the current run
    last_id = self.array_ids.get(key, None)  # check id of pre-allocated array in the previous run of this key
    if last_id is not None and last_id != current_id:
        rfft_obj.update_arrays(real_arr, complex_arr)
        irfft_obj.update_arrays(complex_arr, real_arr)
    self.array_ids[key] = current_id
```

**Check performance**
The suggestion above is included in the `challenger_sdp.py` module I developed locally. Note that the baseline is [the new version of pyfftw_sdp in this PR](https://github.com/stumpy-dev/sliding_dot_product/blob/challenge_pyfftw/sdp/pyfftw_sdp.py). To have a better picture, I've included the original pyfftw_sdp (of `main` branch), and [the INITIAL proposal provided in this PR](https://github.com/stumpy-dev/sliding_dot_product/pull/25/commits/c8074a285c7dba6fa1e49614a99928b602bf85cc).


<img width="1354" height="2840" alt="Image" src="https://github.com/user-attachments/assets/481d23cc-ed04-4725-a68a-c7c35ac980db" />


**Observation**
The baseline is [the new version of pyfftw_sdp in this PR](https://github.com/stumpy-dev/sliding_dot_product/blob/challenge_pyfftw/sdp/pyfftw_sdp.py). The `challenger_sdp.py` (green) has the aforementioned enhancement, and it shows a minor enhancement boost. Its performance is now slightly more than the original `pyfftw_sdp` (orange).

NimaSarajpoor replied:
While there is a performance gain for shorter arrays, the added code complexity may not be justified at this stage. We can merge this PR and revisit these optimizations later if we identify a specific need for better short-array performance.

seanlaw replied:
I agree. Let's merge the simplest, cleanest version that gets us 90%+ of the way there.

NimaSarajpoor reviewed (commented):
@seanlaw 
I think this PR is ready. I discovered a small opportunity for improvement and left a comment. Wanted to know your input before I merge this PR.

NimaSarajpoor commented:
Thank you @seanlaw for all the comments and feedbacks.
